source/common/json/json_streamer.h

```diff
@@ -1,25 +1,119 @@
+  // Constants for JSON delimiters.
+  static constexpr absl::string_view MapBeg = R"({)";
+  static constexpr absl::string_view MapEnd = R"(})";
+  static constexpr absl::string_view ArrayBeg = R"([)";
+  static constexpr absl::string_view ArrayEnd = R"(])";
+  static constexpr absl::string_view Quote = R"(")";
+  static constexpr absl::string_view Comma = R"(,)";
+};
+
+/**
+ * Helper class that provides low level methods to sanitize keys, values, and
+ * delimiters to output buffer or string.
+ * NOTE: Streamer provides high level methods based on the Serializer to construct
+ * complete JSON output. Please use Streamer as priority.
+ * NOTE: If Serializer is preferred please ensure YOU KNOW WHAT YOU ARE DOING.
+ */
+template <class OutputBuffer> class Serializer {
+public:
+  // Constructor with output buffer to write the JSON pieces.
+  Serializer(OutputBuffer& output_buffer) : output_buffer_(output_buffer) {}
+
+  /**
+   * Add raw JSON piece to the output buffer. The value could be delimiters,
+   * or sanitized JSON piece.
+   *
+   * @param value The raw JSON piece to be added.
+   */
+  void addPiece(absl::string_view value) { output_buffer_.addFragments({value}); }
+
+  /**
+   * Add a string value to the JSON buffer. The string value will be sanitized
+   * per JSON rules.
+   *
+   * @param value The string value or key to be sanitized and added.
+   * @param prefix prefix string that will be appended before the value.
+   * Default be `"`.
+   * @param suffix suffix string that will be appended after the value.
+   * Default be `"`.
+   *
+   * NOTE: Both key and string values should use this method to sanitize.
+   */
+  void addString(absl::string_view value, absl::string_view prefix = Constants::Quote,
+                 absl::string_view suffix = Constants::Quote) {
+    // Sanitize the string value and quote it on demand of the caller.
+    absl::string_view sanitized = Json::sanitize(sanitize_buffer_, value);
+    output_buffer_.addFragments({prefix, sanitized, suffix});
+  }
+
+  /**
+   * Add a number value to the raw JSON piece buffer.
+   *
+   * @param value The number value to be added.
+   * TODO(wbpcode): will fmt::format_int provide better performance for integer?
+   */
+  void addNumber(int64_t value) { output_buffer_.addFragments({absl::StrCat(value)}); }
+  void addNumber(uint64_t value) { output_buffer_.addFragments({absl::StrCat(value)}); }
+  void addNumber(double value) {
+    if (std::isnan(value)) {
+      output_buffer_.addFragments({Constants::Null});
+    } else {
+      Buffer::Util::serializeDouble(value, output_buffer_);
+    }
+  }
+
+  /**
+   * Add a bool value to the raw JSON piece buffer.
+   *
+   * @param value The bool value to be added.
+   */
+  void addBool(bool value) {
+    output_buffer_.addFragments({value ? Constants::True : Constants::False});
+  }
+
+  /**
+   * Add a null value to the raw JSON piece buffer.
+   */
+  void addNull() { output_buffer_.addFragments({Constants::Null}); }
```

Comment:
These are completely same with original impl except the `output_buffer_` is templated.